synapse/handlers/pagination.py

```diff
@@ -72,13 +72,138 @@ def __init__(self, hs):
+                )
+                defer.returnValue(False)
+
+        defer.returnValue(False)
+
+    @defer.inlineCallbacks
+    def purge_history_for_rooms_in_range(self, min_ts, max_ts):
+        """Purge outdated events from every room which maximum retention period is within
+        ]min_ts ; max_ts]. If the configured maximum period in the server's configuration
+        is within this range, also targets rooms which don't have a retention policy.
+
+        Args:
+            min_ts (int): Timestamp in milliseconds that define the lower limit of the
+                range to handle.
+            max_ts (int): Timestamp in milliseconds that define the higher limit of the
+                range to handle.
+        """
+        # If a room lacks a max_lifetime, we consider it equal to one defined in the
+        # server's configuration, therefore we include these rooms if the server's
+        # config's max_lifetime is in the provided range.
+        include_null = (min_ts < self.config.retention_max_lifetime <= max_ts)
+        rooms = yield self.store.get_rooms_for_retention_period_in_range(
+            min_ts, max_ts, include_null
+        )
+
+        for room in rooms:
+            room_id = room["room_id"]
+
+            max_lifetime = room["max_lifetime"]
+
+            if max_lifetime is None:
+                max_lifetime = self.config.retention_max_lifetime
+
+            # Figure out what token we should start purging at.
+            ts = self.clock.time_msec() - max_lifetime
+
+            stream_ordering = (
+                yield self.store.find_first_stream_ordering_after_ts(ts)
+            )
+
+            r = (
+                yield self.store.get_room_event_after_stream_ordering(
+                    room_id, stream_ordering,
+                )
+            )
+            if not r:
+                logger.warning(
+                    "[purge] purging events not possible: No event found "
+                    "(ts %i => stream_ordering %i)",
+                    ts, stream_ordering,
+                )
+                continue
+
+            (stream, topo, _event_id) = r
+            token = "t%d-%d" % (topo, stream)
+
+            if room_id in self._purges_in_progress_by_room:
+                logger.warning(
+                    "[purge] not purging room %s as there's an ongoing purge running"
+                    " for this room",
+                    room_id,
+                )
+                continue
+
+            purge_id = random_string(16)
+
+            self._purges_by_id[purge_id] = PurgeStatus()
+
+            logger.info(
+                "Starting purging events in room %s (purge_id %s)" % (room_id, purge_id)
+            )
+
+            # We want to purge everything, including local events, and to run the purge in
+            # the background so that it's not blocking any other operation apart from
+            # other purges in the same room.
+            run_in_background(
```

Comment:
can you use `run_as_background_process` for this, so that we don't lose the metrics for it?